### PR TITLE
storage-client: add param to disable shard finalization

### PR DESCRIPTION
**Pull request by sploiselle**
Allow disabling shard finalization.

### Motivation

This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql/src/session/vars.rs
Comment (pH14) [inline src/sql/src/session/vars.rs:967]: `Whether to all the `

File: src/sql/src/session/vars.rs
Comment (pH14) [inline src/sql/src/session/vars.rs:963]: nit: copypasta

File: src/sql/src/session/vars.rs
Comment (pH14) [inline src/sql/src/session/vars.rs:2217]: nit: copypasta

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:963]: copy🍝

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:967]:
```suggestion
    description: "Whether to alllow the storage client to finalize shards (Materialize).",
```

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:965]: I think we normally name these boolean vars with an `enable_` prefix?

```suggestion
    name: UncasedStr::new("enable_storage_shard_finalization"),
```

File: src/storage-client/src/controller.rs
Comment (benesch) [inline src/storage-client/src/controller.rs:2300]: Thoughts on logging some informational messages here?

```suggestion
                if self.state.config.finalize_shards {
                    info!("triggering shard finalization due to dropped storage object");
                    self.finalize_shards().await;
                } else {
                    info!("not triggering shard finalization due to dropped storage object because enable_storage_shard_finalization parameter is false")
                }
```

**Review by benesch — COMMENTED** [review]
Thanks for turning this around so quickly! Going to mark this as a release blocker for tomorrow.

File: src/sql/src/session/vars.rs
Comment (pH14) [inline src/sql/src/session/vars.rs:965]: I can't find the link right now, but I think the style suggestion is to phrase as `enable_` and default false / override true

File: src/sql/src/session/vars.rs
Comment (pH14) [reply src/sql/src/session/vars.rs:963]: lol nevermind me, Nikhil is too fast

File: src/storage-client/src/controller.rs
Comment (pH14) [reply src/storage-client/src/controller.rs:2300]: related thought: a metric gauge for the # of awaiting-finalization shards would be handy

**Comment by benesch** [comment]
Deferring review to the persist/storage team.

**Comment by sploiselle** [comment]
@jkosh44 I left this to `true` because this isn't really a "feature" flag--I'd used a different name originally to avoid the confusion, but that both you and Nikhil suggested the same name instead suggested there was room for improvement. I think there's time to sort this out and the little bit of thrashing in this code isn't a huge deal, just need to know we need to explicitly set this to false.

**Comment by benesch** [comment]
Defaulting to `true` makes sense to me, as it matches the current behavior of the code. We can override to `false` for the affected customer environment.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
> Deferring official approval to the persist/storage team. Thanks again, @sploiselle!

Approving as I've thought harder about this PR and it seems 100% safe to me. Per discussion on Slack, we're going to wait until 7pm to see about getting a review from @pH14 and otherwise proceed with the merge and backport.

**Pull request merged**
